torch/csrc/jit/python/init.cpp

```diff
@@ -897,13 +898,65 @@ void initJITBindings(PyObject* module) {
             docstring << "  " << op->schema() << "\n";
           }
 
-          return py::cpp_function(
-              [operations](py::args args, py::kwargs kwargs) {
+          // To support __torch_function__ override mechanism, `func` needs to
+          // provide a pointer, which points to the original method, i.e.,
+          // itself. py::cpp_function doesn't support specifying the function
+          // signature without real implementation, so we we use a trick here.
+          // First, we create a function holder pointing to an empty
+          // cpp_function. Then we use a reference as the alias of the empty
+          // function. Then we define the real returned function, and update
+          // the value of the function pointed by the holder using the alias.
+          // With this trick, we are able to reference self inside the returned
+          // function.
```

Comment:
I'm concerned about reference cycle here. `func_holder` holds an owning reference to the (refcounted) `cpp_function` and you then transfer ownership of this unique reference to the `cpp_function`. So `cpp_function` has a owning reference to itself, reference cycle.

What you should probably do instead maintain a non-owning to cpp function inside the cpp function. You are guaranteed that it will be live when you use it in the function because, how else could you have called the function; only when you pass it into a function call do you increase refcount.

cc @dzhulgakov

Reply:
okay, i will just use py::import to get the original method.

Reply:
good catch, we could you `py::handle` (the unowning reference). But py::import works too since perf doesn't matter when we trigger `__torchfunction__`